yt_dlp/extractor/vrsquare.py

```diff
@@ -0,0 +1,171 @@
+            'id': video_id,
+            'title': self._html_search_meta(['og:title', 'twitter:title'], webpage),
+            'description': self._html_search_meta('description', webpage),
+            'formats': self._extract_m3u8_formats(traverse_obj(webApi, (
+                'urls', ..., 'url', any)), video_id, 'mp4', fatal=False),
+            'thumbnail': self._html_search_meta('og:image', webpage),
+            **traverse_obj(webpage, {
+                'duration': ({find_element(cls='layout-product-data-time')}, {parse_duration}),
+                'tags': ({find_elements(cls='search-tag')}, ..., {clean_html}),
+            }),
+        }
+
+
+class VRSQUAREPlaylistIE(VRSQUAREIE):
+    IE_NAME = 'vrsquare:playlist'
+
+    _VALID_URL = r'https?://livr\.jp/(?P<type>category|channel|headline|web-search)/?(?P<id>(?!top\b)[\w-]+)?'
+    _TESTS = [{
+        'url': 'https://livr.jp/category/C133936275',
+        'info_dict': {
+            'id': 'C133936275',
+            'title': 'そこ曲がったら、櫻坂？VR',
+        },
+        'playlist_mincount': 308,
+    }, {
+        'url': 'https://livr.jp/channel/H372648599',
+        'info_dict': {
+            'id': 'H372648599',
+            'title': 'AKB48＋チャンネル',
+        },
+        'playlist_mincount': 499,
+    }, {
+        'url': 'https://livr.jp/headline/A296449604',
+        'info_dict': {
+            'id': 'A296449604',
+            'title': 'AKB48 アフターVR',
+        },
+        'playlist_mincount': 22,
+    }, {
+        'url': 'https://livr.jp/web-search?w=%23%E5%B0%8F%E6%A0%97%E6%9C%89%E4%BB%A5',
+        'info_dict': {
+            'id': '#小栗有以',
+            'title': '#小栗有以',
+        },
+        'playlist_mincount': 70,
+    }]
+
+    def _fetch_vids(self, source, keys=()):
+        return [self.url_result(
+            f'{self._BASE_URL}/contents/{x.removeprefix("/contents/")}', VRSQUAREIE)
+            for x in traverse_obj(source, (
+                *keys, {find_elements(cls='video', html=True)},
+                ..., {extract_attributes}, 'data-url'))
+        ]
+
+    def _entries(self, webpage, playlist_type, playlist_id, query):
+        if playlist_type in ('category', 'headline'):
+            yield from self._fetch_vids(webpage)
+        else:
+            endpoint = {
+                'channel': f'channel/{playlist_id}',
+                'web-search': 'web-search',
+            }[playlist_type]
+            for page in itertools.count(1):
+                ajax = self._download_json(
+                    f'{self._BASE_URL}/ajax/{endpoint}', playlist_id,
+                    query=filter_dict({'p': page, 'w': query}),
+                )
+                yield from self._fetch_vids(ajax, ('contents_render_list', ...))
+                if not any(ajax.get(k) for k in ('has_next', 'hasNext')):
+                    break
+
+    def _real_extract(self, url):
+        playlist_type, playlist_id = self._match_valid_url(url).groups()
+        query = parse_qs(url).get('w', [None])[0]
```

Comment:
I think I would move some of the shared logic in to a PlaylistBase class and then have a separate extractors for web-serach, channel, and headline/category.